### PR TITLE
save before copy if notebook is dirty

### LINKED ISSUE #986: "Make a copy" menu item loses unsaved changes
Sometimes I find myself making changes to a notebook, then realise that I should keep my original notebook for comparison reasons.

So I use the "Make a copy" option to save a copy of the notebook with my changes (and go back to undo them from my original file). However, this make a copy of the current *notebook state on disk, neglecting any unsaved changes*.

I appreciate that this may be subjective, but this behaviour seems quite counter-intuitive to me. My expectation of this option is that it should operate more like "Save As...", so that the current contents of the notebook gets saved to a new filename.

### PR DESCRIPTION
**Pull request by minrk**
in copy_notebook, rather than relying on async: false in event handlers, which is ignored.

closes #986

### CODE REVIEW

File: notebook/static/notebook/js/notebook.js [hunk 1]
```diff
@@ -2578,23 +2578,32 @@ define(function (require) {
...
-            }
-        );
+        var p;
+        if (this.dirty) {
+            p = this.save_notebook();
+        } else {
+            p = Promise.resolve();
+        }
+        p.then(function () {
+            that.contents.copy(that.notebook_path, parent).then(
```
Comment (jdfreder) [inline notebook/static/notebook/js/notebook.js]: If you add a `return` both here and the line above then `copy_notebook` notebook method will return an appropriate promise, which is useful for callers of the method.

**Comment by Carreau** [comment]
relaunched tests.

File: notebook/static/notebook/js/notebook.js (hunk 1 above)
Comment (minrk) [inline notebook/static/notebook/js/notebook.js]: Good call. Added.

**Comment by Carreau** [comment]
Relaunched test. 
Candidate for backport.

**Pull request merged**